feat: enable native bb

charlielye opened the pull request:
This PR enables native bb code behind bb.js. It leverages the backends and ipc mechanisms from previous work.

Important points of note:
* `Barretenberg` - Use for big jobs. It has an async api. It will use a Unix Domain Socket (UDS) to speak to bb with msgpack protocol.
* `BarretenbergSync` - Use for small jobs. Hashes and the like. It will use the NAPI module which has a client side IPC stub to speak with bb process using shared memory. WARNING: The io buffer sizes are fixed and are controlled via cli args. Current input buffer size is set to 2mb and response to default 1mb. This is a little tech heavy, so if it does introduce some fundamental issue, we can simply comment out the line in bb.js that tries to init shm first, and go back to wasm (what we had before).
* We leverage **dependency injection** r.e. Barretenberg. i.e. it's injected into `UltraHonk(Verifier)Backend` and `AztecClientBackend`. We need to do more dependency injection...
* Note that if you want logs from BarretenbergSync when it's in SHM mode, the logs will be written to a /tmp file, the path of which will be given to you on the TS logs when it inits. That's because as a synchronous system node can't print the logs for you until it returns. So in the event there's really weird hangs or the like (shouldn't happen unless there's bugs in the IPC code or cbind codegen tool), you can use the log file. Regular errors will be returned as an api response and TS will throw it on.

More broadly the work in this PR:
* Added prettier to bbjs-test.
* Seperates api_msgpack.hpp into cpp/hpp.
* Msgpack api sends back ErrorResponse to caller.
* Add ring buffer size cli args for shm ipc.
* The IPC api is updated to work in "message frames". SHM is zero copy in the sense that it parses msgpack data directly out the memory written to by node napi module. UDS IPC has it's own internal buffer which can grow unbounded.
* A whole bunch of sol files in `barretenberg/sol` got reformatted, I guess due to prior foundry upgrade. Don't really know where they came from but they kept coming back after reverting so. They're in.
* We get rid of the `kill_wrapper.sh` script in favour on linux using PDEATH (parent death) signal from kernel on linux, and monitoring a kernel event on macos. Cleaner process cleanup. We need such a solution to take burden of needing to destroy the Barretenberg(Sync) classes on the developer, which otherwise causes node to not exit. It's good practice to do proper cleanup but yeah... Let's be honest, it's going to confuse people because we also have lazy initialisation of bb (another can of worms I might open up again later).
* Revert all the BB_WASM_PATH nastiness in favour of just checking for that env var override in one place.
* Remove the last of the old cbind logic from bb.js (it was only used by the poseidon bench).
* Make the ivc-integration browser test great again.

Still TODO:
* There's some chonk calls that still use the cli directly. Verification comes to mind.
* Consider making "native rollup prover" use bb.js. Currently still calls binary. Think we should do this. Even if we need parallelisation you can still init multiple Barretenberg's (and associated bb process).
* Consider making "avm proving" go over the msgpack api.
* Stepping stone, extend bb.js ts api to get the path to your arches bb binary, we can then use that instead of having to provide a path to bb external to the project (helps escape docker container).

ludamad commented on barretenberg/cpp/scripts/ci_benchmark_browser_memory.sh:
```diff
@@ -14,12 +14,6 @@ export PORT=9339
 
-# Run server.
-yarn --cwd ../acir_tests/browser-test-app serve &
-server_pid=$!
```
Why remove this?

charlielye replied:
We now launch the server from within the test itself.

ludamad commented on barretenberg/cpp/src/barretenberg/api/api_msgpack.cpp:
```diff
@@ -0,0 +1,342 @@
+
+        // Deserialize the msgpack buffer
+        // The buffer should contain a tuple of arguments (array) matching the bbapi function signature.
+        // Since bbapi(Command) takes one argument, we expect a 1-element array containing the Command.
```
Why use an array at all if we are doing a 1-element array?

ludamad replied:
(but definitely a follow-up)

charlielye replied:
It's the cbind abstraction. It allows for calling a function with any number of arguments. Just happens that bbapi just has 1 argument. I think it was this or have two different paths in TS code. Maybe obsolete now.

ludamad commented on barretenberg/cpp/src/barretenberg/api/api_msgpack.cpp:
```diff
@@ -0,0 +1,342 @@
+        return execute_msgpack_ipc_server(std::move(server));
+    }
+#else
+    (void)max_clients;
```
`[[maybe_unused]]` annotation is idiomatic

ludamad commented on barretenberg/cpp/src/barretenberg/ipc/shm/spsc_shm.cpp:
```diff
@@ -307,21 +308,33 @@ void* SpscShm::peek(size_t* n)
-                }
-                continue; // Try again from wrapped position
+        // If we don't have enough bytes to even read a padding marker, treat as implicit padding
+        // This happens when the producer wrapped and left < 4 bytes at the end
```
nit: comment about what 'wrapped' means here would be nice

ludamad commented on barretenberg/cpp/src/barretenberg/ipc/shm_server.hpp:
```diff
@@ -102,55 +108,74 @@ class ShmServer : public IpcServer {
-        uint64_t timeout_us = timeout_ns > 0 ? timeout_ns / 1000 : 100000; // Default 100ms
-        return consumer_->wait_for_data(static_cast<uint32_t>(timeout_us));
+        // Pass spin time directly (don't convert ns to us - underlying API expects ns)
+        return consumer_->wait_for_data(static_cast<uint32_t>(spin_ns / 1000)); // Convert to microseconds for MPSC API
```
Do these comments line up?

ludamad commented on barretenberg/ts/src/bb_backends/node/native_shm.ts:
```diff
@@ -59,22 +64,40 @@ export class BarretenbergNativeShmSyncBackend implements IMsgpackBackendSync {
-    const bbProcess = spawn(findPackageRoot() + '/scripts/kill_wrapper.sh', args, {
-      stdio: ['ignore', 'ignore', 'ignore'],
+    const args = [
+      'msgpack',
```
Unrelated comment to PR: I'm keen to rename msgpack run and msgpack schema to be bbapi run and bbapi schema

ludamad reviewed (approved):
LGTM. Excited for this one.

charlielye committed:
feat: enable native bb

This PR enables native bb code behind bb.js. It leverages the backends and ipc mechanisms from previous work.

Important points of note:
* `Barretenberg` - Use for big jobs. It has an async api. It will use a Unix Domain Socket (UDS) to speak to bb with msgpack protocol.
* `BarretenbergSync` - Use for small jobs. Hashes and the like. It will use the NAPI module which has a client side IPC stub to speak with bb process using shared memory. WARNING: The io buffer sizes are fixed and are controlled via cli args. Current input buffer size is set to 2mb and response to default 1mb. This is a little tech heavy, so if it does introduce some fundamental issue, we can simply comment out the line in bb.js that tries to init shm first, and go back to wasm (what we had before).
* We leverage **dependency injection** r.e. Barretenberg. i.e. it's injected into `UltraHonk(Verifier)Backend` and `AztecClientBackend`. We need to do more dependency injection...
* Note that if you want logs from BarretenbergSync when it's in SHM mode, the logs will be written to a /tmp file, the path of which will be given to you on the TS logs when it inits. That's because as a synchronous system node can't print the logs for you until it returns. So in the event there's really weird hangs or the like (shouldn't happen unless there's bugs in the IPC code or cbind codegen tool), you can use the log file. Regular errors will be returned as an api response and TS will throw it on.

More broadly the work in this PR:
* Added prettier to bbjs-test.
* Seperates api_msgpack.hpp into cpp/hpp.
* Msgpack api sends back ErrorResponse to caller.
* Add ring buffer size cli args for shm ipc.
* The IPC api is updated to work in "message frames". SHM is zero copy in the sense that it parses msgpack data directly out the memory written to by node napi module. UDS IPC has it's own internal buffer which can grow unbounded.
* A whole bunch of sol files in `barretenberg/sol` got reformatted, I guess due to prior foundry upgrade. Don't really know where they came from but they kept coming back after reverting so. They're in.
* We get rid of the `kill_wrapper.sh` script in favour on linux using PDEATH (parent death) signal from kernel on linux, and monitoring a kernel event on macos. Cleaner process cleanup. We need such a solution to take burden of needing to destroy the Barretenberg(Sync) classes on the developer, which otherwise causes node to not exit. It's good practice to do proper cleanup but yeah... Let's be honest, it's going to confuse people because we also have lazy initialisation of bb (another can of worms I might open up again later).
* Revert all the BB_WASM_PATH nastiness in favour of just checking for that env var override in one place.
* Remove the last of the old cbind logic from bb.js (it was only used by the poseidon bench).
* Make the ivc-integration browser test great again.

Still TODO:
* There's some chonk calls that still use the cli directly. Verification comes to mind.
* Consider making "native rollup prover" use bb.js. Currently still calls binary. Think we should do this. Even if we need parallelisation you can still init multiple Barretenberg's (and associated bb process).
* Consider making "avm proving" go over the msgpack api.
* Stepping stone, extend bb.js ts api to get the path to your arches bb binary, we can then use that instead of having to provide a path to bb external to the project (helps escape docker container).
